src/py/mat3ra/made/tools/analyze/coordination.py

```diff
@@ -0,0 +1,499 @@
+    voronoi_nn = PymatgenVoronoiNN(
+        tol=tolerance,
+        cutoff=cutoff,
+        weight="solid_angle",
+        extra_nn_info=False,
+        compute_adj_neighbors=True,
+    )
+    coordinates = material.basis.coordinates
+    site_index = coordinates.get_element_id_by_value(coordinate)
+    remove_dummy_atom = False
+    if site_index is None:
+        structure.append("X", coordinate, validate_proximity=False)
+        site_index = len(structure.sites) - 1
+        remove_dummy_atom = True
+    try:
+        neighbors = voronoi_nn.get_nn_info(structure, site_index)
+    except ValueError:
+        return None
+    neighboring_atoms_pymatgen_ids = [n["site_index"] for n in neighbors]
+    if remove_dummy_atom:
+        structure.remove_sites([-1])
+
+    all_coordinates = material.basis.coordinates
+    all_coordinates.filter_by_indices(neighboring_atoms_pymatgen_ids)
+    return all_coordinates.ids
+
+
+def get_undercoordinated_atom_indices(material: Material, cutoff: float, coordination_threshold: int) -> List[int]:
+    """
+    Identify undercoordinated atoms based on the coordination threshold (inclusive).
+
+    Args:
+        material (Material): The material object.
+        cutoff (float): The cutoff radius for identifying neighbors.
+        coordination_threshold (int): The coordination number threshold for an atom to be considered undercoordinated.
+
+    Returns:
+        List[int]: List of indices of undercoordinated atoms.
+    """
+    coordination_numbers = get_coordination_numbers(material, cutoff)
+    return [idx for idx, number in coordination_numbers.items() if number <= coordination_threshold]
+
+
+def get_unique_coordination_numbers(material: Material, cutoff: float) -> List[int]:
+    """
+    Get the unique coordination numbers for all atoms in the material.
+
+    Args:
+        material (Material): The material object.
+        cutoff (float): The cutoff radius for identifying neighbors.
+
+    Returns:
+        List[int]: A list of unique coordination numbers present in the material.
+    """
+    coordination_numbers = get_coordination_numbers(material, cutoff)
+    return sorted(list(set(coordination_numbers.values())))
+
+
+def find_unique_bond_directions(material: Material, angle_tolerance: float = 0.1) -> Dict[str, List[np.ndarray]]:
+    """
+    Find unique bond templates for each element type.
+
+    Args:
+        atom_vectors (List[List[np.ndarray]]): List of bond vectors for each atom.
+        atom_elements (List[str]): List of chemical elements for each atom.
+        angle_tolerance (float): Tolerance for comparing angles between bond vectors.
+
+    Returns:
+        Dict[str, List[np.ndarray]]: Dictionary mapping element types to unique bond templates.
+    """
+    element_templates = {}
+
+    for element in set(material.basis.elements.values):
+        element_templates[element] = find_bond_directions_for_element(material, element, angle_tolerance)
```

Comment:
We need to define a structure that will hold this - I'm suggesting a flat list with {element: "Si", bond_direction: BondDirection()} as members